### PR TITLE
fix: Multitenant apis

### PR DESCRIPTION
**Pull request by sattvikc**
Regarding multi-tenancy recipe:

- Added 2 APIs for user association with tenant.
- Updated existing APIs to work with existing core changes.

### CODE REVIEW

File: api_spec.yaml [hunk 1]
```diff
@@ -3454,59 +4017,13 @@ paths:
...
-          $ref: '#/components/responses/401'
-
-        '404':
-          $ref: '#/components/responses/404'
-
-        '500':
-          $ref: '#/components/responses/500'
-
   /recipe/multitenancy/config/thirdparty:
     put:
```
Comment (rishabhpoddar) [inline api_spec.yaml:4021]: remove the additional config stuff and keep it only for apple, google workspaces and boxyhq

File: api_spec.yaml [hunk 2]
```diff
@@ -3454,59 +4017,13 @@ paths:
...
-        '401':
-          $ref: '#/components/responses/401'
-
-        '404':
-          $ref: '#/components/responses/404'
-
-        '500':
-          $ref: '#/components/responses/500'
-
   /recipe/multitenancy/config/thirdparty:
```
Comment (rishabhpoddar) [inline api_spec.yaml:4020]: we can remove /recipe/multitenancy/config/thirdparty/list API

File: api_spec.yaml [hunk 3]
```diff
@@ -3736,6 +4247,119 @@ paths:
...
           $ref: '#/components/responses/500'
 
+  /recipe/multitenancy/tenant/user:
+    post:
+      tags:
+        - Multitenancy Recipe
+      description: |
+        Add user's association with the tenant.
+
+        Note: To associate a user to a tenant, the tenant must be in the same user pool
```
Comment (rishabhpoddar) [inline api_spec.yaml:4257]:
```suggestion
        Note: To associate a user to a tenant, the tenant must be in the same database
```

File: api_spec.yaml [hunk 4]
```diff
@@ -3736,6 +4247,119 @@ paths:
         '500':
           $ref: '#/components/responses/500'
 
+  /recipe/multitenancy/tenant/user:
+    post:
+      tags:
+        - Multitenancy Recipe
+      description: |
+        Add user's association with the tenant.
+
```
Comment (rishabhpoddar) [inline api_spec.yaml:4256]: add to the description that the userId will be added to the tenant id based on the prefix of the url

File: api_spec.yaml [hunk 5]
```diff
@@ -3736,6 +4247,119 @@ paths:
...
+                - type: object
+                  required:
+                    - status
+                    - wasAlreadyAssociated
+                  properties:
+                    status:
+                      $ref: '#/components/schemas/statusOK'
+                    wasAlreadyAssociated:
+                      type: boolean
+                - $ref: '#/components/schemas/unknownUserIdResponse'
```
Comment (rishabhpoddar) [inline api_spec.yaml:4291]: there are other error states as well:
- email already exists
- third party id + third party userid already exists
- phone number already exists

File: api_spec.yaml [hunk 6]
```diff
@@ -3736,6 +4247,119 @@ paths:
...
+  /recipe/multitenancy/tenant/user:
+    post:
+      tags:
+        - Multitenancy Recipe
+      description: |
+        Add user's association with the tenant.
+
+        Note: To associate a user to a tenant, the tenant must be in the same user pool
+        and user must already exist in the app
+
```
Comment (rishabhpoddar) [inline api_spec.yaml:4259]: the API for getUser in all the recipes will also return the list of tenants for that user.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: api_spec.yaml [hunk 7]
```diff
@@ -3247,56 +3248,592 @@ paths:
...
+        This request must originate from public app and public tenant on the base
+        connection uri domain.
+
+        If creating a new connection uri domain, only the login methods set to true will be enabled and rest will be disabled by default.
+
+        If updating an existing connection uri domain, core will keep the existing state of login methods and only update the ones that are specified in the request body.
+
+        Note: The core config must contain a unique connection to the storage layer, because sharing of database between 2 different connection uri domains is not allowed.
+
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3269]: Done

File: api_spec.yaml [hunk 8]
```diff
@@ -3247,56 +3248,592 @@ paths:
...
+      tags:
+        - Multitenancy Recipe
+      description: |
+        Delete a connection uri domain.
+
+        Deletion of base connection uri domain is not allowed. Instead use the `/recipe/multitenancy/connectionuridomain PUT` api to disable all the recipes.
+
+        The request must originate from base connection uri domain and public app and tenant.
+
+        Note: There should be no apps or tenants belonging to the connection uri domain to be able to delete it.
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3335]: Done

File: api_spec.yaml [hunk 9]
```diff
@@ -3247,56 +3248,592 @@ paths:
...
+            application/json:
+              schema:
+                type: object
+                required:
+                  - status
+                  - connectionUriDomainExisted
+                properties:
+                  status:
+                    $ref: '#/components/schemas/statusOK'
+                  tenantExisted:
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3366]: Done

File: api_spec.yaml [hunk 10]
```diff
@@ -3247,56 +3248,592 @@ paths:
...
+        '401':
+          $ref: '#/components/responses/401'
+
+        '404':
+          $ref: '#/components/responses/404'
+
+        '500':
+          $ref: '#/components/responses/500'
+
+  /recipe/multitenancy/connectionuridomain/config:
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3767]: Removed

File: api_spec.yaml [hunk 11]
```diff
@@ -3312,47 +3849,75 @@ paths:
         '500':
           $ref: '#/components/responses/500'
 
-  /recipe/multitenancy/tenant/remove:
-    post:
+  /recipe/multitenancy/app/config:
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3852]: Removed

File: api_spec.yaml [hunk 12]
```diff
@@ -3312,47 +3849,75 @@ paths:
...
+                        enabled:
+                          type: boolean
+                - type: object
+                  required:
+                    - status
+                  properties:
+                    status:
+                      $ref: '#/components/schemas/statusTenantDoesNotExistError'
+
         '400':
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3921]: Done

File: api_spec.yaml (hunk 3 above)
Comment (sattvikc) [reply api_spec.yaml:4257]: Done

File: api_spec.yaml (hunk 4 above)
Comment (sattvikc) [reply api_spec.yaml:4256]: Done

File: api_spec.yaml (hunk 5 above)
Comment (sattvikc) [reply api_spec.yaml:4291]: DONE

File: api_spec.yaml [hunk 13]
```diff
@@ -3312,47 +3849,75 @@ paths:
...
         - Multitenancy Recipe
       description: |
-        Delete a tenant.
+        Get an app config.
 
-        If `tenantId` is not equal to `defaultTenantId`, SuperTokens subscription license key is required.
+        SuperTokens subscription license key is required.
 
-        Deletion of `defaultTenantId` is not allowed. Instead use the `/recipe/multitenancy/tenant PUT` api to disable all the recipes.
+        Returns recipes with their enabled flag and recipe specific configs saved in core
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply api_spec.yaml:3862]: Done

File: api_spec.yaml (hunk 2 above)
Comment (sattvikc) [reply api_spec.yaml:4020]: Done

File: api_spec.yaml (hunk 1 above)
Comment (sattvikc) [reply api_spec.yaml:4021]: Done

**Pull request merged**